### PR TITLE
Feature/better passing logic

### PR DESCRIPTION
**Pull request by RobZelluf**
To do:
- Using PassScore to determine robot to pass to
- Implement quickPass (take into account how quick the robot could shoot the ball there)
- Come up with something better to penalize passing backwards

### CODE REVIEW

File: roboteam_ai/test/helpers/FieldHelper.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    field.bottom_right_penalty_stretch.end = Vector2{field.field_length / 2, -defenseAreaWidth / 2};
+
+    field.top_left_penalty_stretch.begin = Vector2{-field.field_length / 2 - defenseAreaDepth, defenseAreaWidth / 2};
+    field.top_left_penalty_stretch.end = Vector2{-field.field_length / 2, defenseAreaWidth / 2};
+    field.bottom_left_penalty_stretch.begin = field.top_left_penalty_stretch.begin;
+    field.bottom_left_penalty_stretch.end = Vector2{-field.field_length / 2, -defenseAreaWidth / 2};
+    return field;
+}
+
+roboteam_msgs::GeometryFieldSize
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/test/helpers/FieldHelper.cpp]: I want side effects plz

File: roboteam_ai/test/helpers/FieldHelper.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    field.field_length = field_length;
+    field.field_width = field_width;
+    field.goal_width = goal_width;
+
+    field = addDefenseAreas(field, defense_area_width, defense_area_length);
+    field = addCenterArc(field, center_circle_radius);
+    return field;
+}
+
+    roboteam_msgs::GeometryFieldSize
```
Comment (ThijsLuttikhuis) [inline roboteam_ai/test/helpers/FieldHelper.cpp]: looks weird

Comment (ThijsLuttikhuis) [reply roboteam_ai/test/helpers/FieldHelper.cpp]: no u

Comment (ThijsLuttikhuis) [reply roboteam_ai/test/helpers/FieldHelper.cpp]: n/o _Uu

Comment (RobZelluf) [reply roboteam_ai/test/helpers/FieldHelper.cpp]: No, u

File: roboteam_ai/test/helpers/FieldHelper.cpp (hunk 1 above)
Comment (RobZelluf) [reply roboteam_ai/test/helpers/FieldHelper.cpp]: Yes

Comment (RobZelluf) [reply roboteam_ai/test/helpers/FieldHelper.cpp]: Fixed

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 3]
```diff
@@ -10,10 +11,11 @@ namespace coach {
 
 const double CoachHeuristics::MAX_DISTANCE_FROM_BALL = 6.0;
 const double CoachHeuristics::CLOSE_TO_GOAL_WEIGHT = -0.1;
-const double CoachHeuristics::SHOT_AT_GOAL_WEIGHT = -2.0;
-const double CoachHeuristics::PASS_LINE_WEIGHT = -2.0;
-const double CoachHeuristics::DISTANCE_TO_OPPONENTS_WEIGHT = -0.5;
+const double CoachHeuristics::SHOT_AT_GOAL_WEIGHT = -3.0;
+const double CoachHeuristics::PASS_LINE_WEIGHT = -3.0;
+const double CoachHeuristics::DISTANCE_TO_OPPONENTS_WEIGHT = -3;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: make it 3.0

Comment (mrlukasbos) [reply roboteam_ai/src/coach/CoachHeuristics.cpp]: -3.0 i mean

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 4]
```diff
@@ -25,16 +27,26 @@ double CoachHeuristics::calculateCloseToGoalScore(const Vector2 &position) {
...
 
-    return 1 - exp(SHOT_AT_GOAL_WEIGHT * shortestDistance);
-}
+    }
 
 /// Gives a higher score if the distance between the ball and the positions if free (safe pass line)
-double CoachHeuristics::calculatePassLineScore(const Vector2& position, const WorldData& world) {
-    double shortestDistance = control::ControlUtils::closestEnemyToLineDistance(world.ball.pos, position, world, false);
+double CoachHeuristics::calculatePassLineScore(const Vector2& position, WorldData world) {
+    double smallestAngle = M_PI / 4;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: what magic is this?

File: roboteam_ai/src/coach/CoachHeuristics.cpp [hunk 5]
```diff
@@ -86,11 +80,22 @@ double CoachHeuristics::calculatePassScore(const Vector2 &position) {
     double closeToGoalScore = calculateCloseToGoalScore(position);
     double shotAtGoalScore = calculateShotAtGoalScore(position, world);
     double passLineScore = calculatePassLineScore(position, world);
+    double behindBallScore = calculateBehindBallScore(position, world);
+    double distanceToOpponentScore = calculateDistanceToOpponentsScore(position, world);
 
-    double score = closeToGoalScore + 3 * shotAtGoalScore + 2 * passLineScore;
+    double score = closeToGoalScore + shotAtGoalScore + 2 * passLineScore + 4 * behindBallScore + distanceToOpponentScore;
```
Comment (mrlukasbos) [inline roboteam_ai/src/coach/CoachHeuristics.cpp]: i feel magic numbers

File: roboteam_ai/src/control/ControlUtils.cpp [hunk 6]
```diff
@@ -70,6 +70,18 @@ double ControlUtils::constrainAngle(double angle) {
     return angle - M_PI;
 }
 
+bool ControlUtils::isPointProjectionOnLine(const Vector2 &pointToCheck, const Vector2 &lineBegin,
```
Comment (mrlukasbos) [inline roboteam_ai/src/control/ControlUtils.cpp]: add documentation

Comment (mrlukasbos) [reply roboteam_ai/src/control/ControlUtils.cpp]: and rename it

File: roboteam_ai/src/coach/CoachHeuristics.cpp (hunk 3 above)
Comment (RobZelluf) [reply roboteam_ai/src/coach/CoachHeuristics.cpp]: Fixed

File: roboteam_ai/src/coach/CoachHeuristics.cpp (hunk 4 above)
Comment (RobZelluf) [reply roboteam_ai/src/coach/CoachHeuristics.cpp]: Made it a constant in the header file

**Pull request merged**